### ISSUE TITLE
Support for writing policies in Go

### ISSUE THREAD

**Issue by justinvp** [pulumi/pulumi-policy#230]

**Comment by mountainerd** [comment]
I was linked here from the [Crossguard](https://www.pulumi.com/docs/guides/crossguard/) page and I see this was opened in 2020. Is there any update or timeline for when Go will have support for writing policies?

**Comment by alex-unikorny** [comment]
Any due date? Thanks!!

**Comment by hazcod** [comment]
Peeking in., would love this too.

**Comment by agpenton** [comment]
👍

**Comment by smebyp1** [comment]
+1

**Comment by Graham-Beer** [comment]
Any update on when this might be available in Go?

**Comment by jasbir84** [comment]
+1 from me

**Comment by root-ali** [comment]
Any update? thanks.

**Comment by justinvp** [comment]
There is an old hackathon project to add a Go SDK: https://github.com/pulumi/pulumi-policy/compare/hackweek/go-policy

**Comment by im7mortal** [comment]
Thank you @justinvp!

1. Is the Python implementation equivalent to the TypeScript one? I noticed that it lacks the following definition: [TypeScript Definition](https://github.com/pulumi/pulumi-policy/blob/main/sdk/nodejs/policy/schema.ts#L15-L57). I already have a raw Golang implementation that follows the Python version. When I looked at the go-policy branch, I found this definition again: [Golang Definition](https://github.com/pulumi/pulumi-policy/blob/7b5dd0bae20390532baf7e9590389a02c11e0f49/sdk/go/policy.go#L54-L87). However, I couldn't track it in the Python implementation.

2. Are you open to a remake of the integration tests runner? I created [a Docker environment ](https://github.com/im7mortal/pulumi-policy/commit/c97d33d97db7ff68368d12bcbffb4569b5773bba)to run it in isolation. Here’s what I found:
- It takes around 30 minutes.
- It produces too much log output, mainly due to prints from the testing.Environment run command and diagnostic prints. (probably ok)
- It has become bloated, and I assume it was initially designed for only one runtime (Node.js). Python was added later, and I think it will get worse if Go and, eventually, .NET are added in the same manner.

Additionally, it fails for me for some reason. I didn't find any differences compared to the GitHub Actions configuration. I will investigate it deeper.

I made some simple prototype to keep every runtime isolated during integration tests. I will push it later.

**Comment by justinvp** [comment]
> 1. Is the Python implementation equivalent to the TypeScript one?

Yes. The TypeScript implementation does have some more advanced capabilities for filtering to specific resource types, but they're largely equivalent. For the config schema, Python has [this](https://github.com/pulumi/pulumi-policy/blob/cce44759b90ded3b7aba6b0bfa8f30df1297a11c/sdk/python/lib/pulumi_policy/policy.py#L141), but lacks the full type definitions that are present in TypeScript. I think we could/should extend Python to include type information for this, similar to TypeScript, and the go hackathon branch, using TypedDicts.

> 2. Are you open to a remake of the integration tests runner?

Open to it. Ideally more would run in parallel. For example, matrix it out such that each test ran as its own separate GitHub Action job, each of which would be isolated and run in parallel.

**Comment by im7mortal** [comment]
1. I will first try to ensure that Python matches with TypeScript.
2. I have a version that runs in parallel—10 minutes compared to 30. I also added some flags to exclude non-target language policy runtimes. I still want to improve some workflows.

**Comment by im7mortal** [comment]
I also investigated [hackweek/go-policy](https://github.com/pulumi/pulumi-policy/compare/hackweek/go-policy). 

I couldn't take much from it as it was written (*I assume*) in a hurry. Still, it was useful to see a solution from a different perspective.

Right now, I have implemented a GRPC server that decodes all protobuf objects into types defined in the Pulumi library (mostly from the plugin package). I try to reuse already defined types as much as possible.

I am considering what would be the best library interface.

#### 1

My first version followed the Python approach. It was pretty straightforward:
```golang
type Policy struct {
	apitype.Policy
	ValidateStack []StackValidator
	Validate      []ResourceValidator
	Remediate     ResourceRemediator
}
```
The user would only need to fill in Validate, Validate & Remediate, Remediate, or ValidateStack. If the user mixes them, it will fail at runtime.

We could provide functions like NewStackValidationPolicy, which would create policies with the correct fields.

We could also make these fields private.

#### 2

This version is almost the same, except that we have private config and use the Functional Options pattern.

So it would look like this:

```golang
p := apitype.Policy{} // or we also could send every fields separately 
NewPolicy(p, WithRemediation(f), WithValidation([]f) .... )

If the user mixes WithValidation with WithStackValidation, it will fail at runtime. This approach also looks more Go-like.

```
#### 3

I don't see much use for generics here. I tried playing around with them but couldn't find a simple way to implement an interface.


### Violation handling. 
I was also thinking about whether functions should include a ReportViolation function (more of a "JS style" approach) or be more Go-like and return errors.

In **hackweek/go-policy** the ReportViolation function clearly states that it needs two arguments (though in Python URN is optional), whereas returning an error would require a custom type, and the user would need to remember to fill in both fields. However, when it comes to URNs, we already know the URN (at least in most cases; I’m unsure about StackValidation calls).


## Conclusion 


While writing down these suggestions, I think I came up with an idea:

```golang
type Policy struct {
	apitype.Policy
	validateStack []StackValidator
	validate      []ResourceValidator
	remediate     ResourceRemediator
}
```

We would have functions like:


```golang
NewValidationPolicy(p, []f, WithRemediation(f))
```

At the same time, I want the PolicyPack to be customizable and allow the injection of custom implementations of pulumirpc.Analyzer and plugin.Analyzer. However, with private fields, users won't be able to inject plugin.Analyzer or reuse PolicyPack.

We could define an interface:

```golang
type PolicyI struct{
   Validate()
   Remediate()
   ValidateStack()
}
```

The official implementation could know which method to call since it would have access to the private fields. However, this might actually be a bad idea because it would require the user to implement both Policy and PolicyPack if they want customize it.

**Comment by im7mortal** [comment]
@justinvp I am currently creating a plugin to run Go policies. I would like to know if I am heading in the right direction.

1. I am planning to add another executable to _~/.pulumi/bin_.
2.   I am using `sdk/go/pulumi-language-go` as the base for the new plugin.
3.  The plugin will compile the policy code in a temporary directory.
4. Then it will (1) start a new policy executable, (2) perform the same logic with port writing/reading, and (3) proxy all gRPC calls to the policy executable.


I have already created a prototype and tested it.


Out of curiosity, I considered whether Go plugins might be a good option, but I decided that the compile/gRPC approach is more tested and standardized within the Pulumi ecosystem.

If this is the correct approach, should I move the compile logic from sdk/go/pulumi-language-go into a reusable library, or would it be better not to modify this critical code for now?

**Comment by im7mortal** [comment]
I've been looking into different parts of the code and had the idea of compiling `pulumi-analyzer-policy-go` using `pulumi-language-go`. Unfortunately, there's currently no way to compile a Go program without running it immediately. Otherwise, it would be a good approach to compile Go policies on the fly.

I will proceed with the solution I described earlier.

**Comment by tyler-dunkel** [comment]
@AnuiDolphin assigning to the java/c# epic

**Comment by nghiadaulau** [comment]
I opened #460, a Go SDK for policy packs in this repository. It follows the behavior of the Node.js and Python SDKs:

- enforcement levels, remediation, not-applicable results, severity, and initial config
- an advisory diagnostic for unknown values during preview
- typed helpers such as `ValidateResourceOfType[T]` that decode resource properties into Go structs

A pack runs from a directory with `runtime: go` in `PulumiPolicy.yaml`, or as a prebuilt binary. The PR includes unit tests and an integration test against the Pulumi CLI.

@Frassle, the PR description compares it with `policyx`, and a comment on the PR asks whether you would rather have these changes ported into `policyx` in pulumi/pulumi. Feedback from anyone who needs Go policies is welcome on #460.
